### PR TITLE
Update Apache Lucene to 9.12.0

### PR DESCRIPTION
**Pull request by reta**
### Description
Update Apache Lucene to 9.12.0, see please https://lucene.apache.org/core/9_12_0/changes/Changes.html

### Related Issues
N/A

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
> ❌ Gradle check result for [128f5ef](https://github.com/opensearch-project/OpenSearch/commit/128f5efeda0e183fe3b427a816afc9c147072eed): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/44937/)

Couple of interesting issues to look at:

```
java.lang.WrongThreadException: Attempted access outside owning thread
	at java.base/jdk.internal.foreign.MemorySessionImpl.wrongThread(MemorySessionImpl.java:315) ~[?:?]
	at java.base/jdk.internal.misc.ScopedMemoryAccess$ScopedAccessError.newRuntimeException(ScopedMemoryAccess.java:113) ~[?:?]
	at java.base/jdk.internal.misc.ScopedMemoryAccess.copyMemory(ScopedMemoryAccess.java:131) ~[?:?]
	at java.base/jdk.internal.foreign.AbstractMemorySegmentImpl.copy(AbstractMemorySegmentImpl.java:644) ~[?:?]
	at java.base/java.lang.foreign.MemorySegment.copy(MemorySegment.java:2217) ~[?:?]
	at org.apache.lucene.store.MemorySegmentIndexInput.readBytes(MemorySegmentIndexInput.java:169) ~[lucene-core-9.12.0-snapshot-25253a1.jar:9.12.0-snapshot-25253a1 25253a1a0160d5ca30682108a28ee829ada9ffde - 2024-08-19 16:04:32]
	at org.opensearch.index.store.Store$VerifyingIndexInput.readBytes(Store.java:1627) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at org.apache.lucene.store.DataInput.readBytes(DataInput.java:73) ~[lucene-core-9.12.0-snapshot-25253a1.jar:9.12.0-snapshot-25253a1 25253a1a0160d5ca30682108a28ee829ada9ffde - 2024-08-19 16:04:32]
	at org.opensearch.common.lucene.store.InputStreamIndexInput.read(InputStreamIndexInput.java:91) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at java.base/java.io.FilterInputStream.read(FilterInputStream.java:119) ~[?:?]
	at org.opensearch.index.snapshots.blobstore.RateLimitingInputStream.read(RateLimitingInputStream.java:66) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at java.base/java.io.FilterInputStream.read(FilterInputStream.java:119) ~[?:?]
	at org.opensearch.repositories.blobstore.BlobStoreRepository$10.read(BlobStoreRepository.java:3575) ~[opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
	at com.azure.storage.common.Utility.lambda$convertStreamToByteBuffer$1(Utility.java:279) ~[azure-storage-common-12.25.1.jar:12.25.1]
	at reactor.core.publisher.FluxGenerate$GenerateSubscription.slowPath(FluxGenerate.java:271) ~[reactor-core-3.5.20.jar:3.5.20]
	at reactor.core.publisher.FluxGenerate$GenerateSubscription.request(FluxGenerate.java:213) ~[reactor-core-3.5.20.jar:3.5.20]
	at reactor.core.publisher.FluxSubscribeOn$SubscribeOnSubscriber.requestUpstream(FluxSubscribeOn.java:131) ~[reactor-core-3.5.20.jar:3.5.20]
	at reactor.core.publisher.FluxSubscribeOn$SubscribeOnSubscriber.onSubscribe(FluxSubscribeOn.java:124) ~[reactor-core-3.5.20.jar:3.5.20]
	at reactor.core.publisher.FluxGenerate.subscribe(FluxGenerate.java:85) ~[reactor-core-3.5.20.jar:3.5.20]
	at reactor.core.publisher.FluxDefer.subscribe(FluxDefer.java:54) ~[reactor-core-3.5.20.jar:3.5.20]
	at reactor.core.publisher.FluxSubscribeOn$SubscribeOnSubscriber.run(FluxSubscribeOn.java:194) ~[reactor-core-3.5.20.jar:3.5.20]
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84) ~[reactor-core-3.5.20.jar:3.5.20]
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:37) ~[reactor-core-3.5.20.jar:3.5.20]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) [?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) [?:?]
	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:103) ~[reactor-core-3.5.20.jar:3.5.20]
		at reactor.core.publisher.Mono.block(Mono.java:1712) ~[reactor-core-3.5.20.jar:3.5.20]
		at com.azure.storage.common.implementation.StorageImplUtils.blockWithOptionalTimeout(StorageImplUtils.java:151) ~[azure-storage-common-12.25.1.jar:12.25.1]
		at com.azure.storage.blob.BlobClient.uploadWithResponse(BlobClient.java:337) ~[azure-storage-blob-12.23.0.jar:12.23.0]
		at org.opensearch.repositories.azure.AzureBlobStore.lambda$writeBlob$18(AzureBlobStore.java:365) ~[main/:?]
		at org.opensearch.repositories.azure.SocketAccess.lambda$doPrivilegedVoidException$0(SocketAccess.java:83) ~[main/:?]
		at java.base/java.security.AccessController.doPrivileged(AccessController.java:571) ~[?:?]
		at org.opensearch.repositories.azure.SocketAccess.doPrivilegedVoidException(SocketAccess.java:82) ~[main/:?]
		at org.opensearch.repositories.azure.AzureBlobStore.writeBlob(AzureBlobStore.java:364) ~[main/:?]
		at org.opensearch.repositories.azure.AzureBlobContainer.writeBlob(AzureBlobContainer.java:133) ~[main/:?]
		at org.opensearch.repositories.blobstore.BlobStoreRepository.snapshotFile(BlobStoreRepository.java:3587) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
		at org.opensearch.repositories.blobstore.BlobStoreRepository.lambda$executeOneFileSnapshot$63(BlobStoreRepository.java:3003) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
		at org.opensearch.action.ActionRunnable$2.doRun(ActionRunnable.java:89) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
		at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:982) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
		at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) [?:?]
		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) [?:?]
		at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
```

File: server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -4465,7 +4465,7 @@ private void snapshotFile(
     ) throws IOException {
         final BlobContainer shardContainer = shardContainer(indexId, shardId);
         final String file = fileInfo.physicalName();
-        try (IndexInput indexInput = store.openVerifyingInput(file, IOContext.READONCE, fileInfo.metadata())) {
+        try (IndexInput indexInput = store.openVerifyingInput(file, IOContext.READ, fileInfo.metadata())) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java:4468]: @mch2 seems like Azure Repository tests are back on track with this change

File: server/src/main/java/org/opensearch/indices/replication/SegmentFileTransferHandler.java [hunk 2]
```diff
@@ -97,20 +98,26 @@ public MultiChunkTransfer<StoreFileMetadata, FileChunk> createTransfer(
         return new MultiChunkTransfer<>(logger, threadPool.getThreadContext(), listener, maxConcurrentFileChunks, Arrays.asList(files)) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/indices/replication/SegmentFileTransferHandler.java:98]: @mch2 did a few changes here to fix the tests, could you please take a look? thank you!

File: server/src/internalClusterTest/java/org/opensearch/action/admin/cluster/node/tasks/ConcurrentSearchTasksIT.java [hunk 3]
```diff
@@ -110,7 +110,7 @@ public void testConcurrentSearchTaskTracking() {
             Map<Long, List<ThreadResourceInfo>> threadStats = getThreadStats(SearchAction.NAME + "[*]", taskInfo.getTaskId());
             // Concurrent search forks each slice of 5 segments to different thread (see please
             // https://github.com/apache/lucene/issues/12498)
-            assertEquals((int) Math.ceil(getSegmentCount(INDEX_NAME) / 5.0) + 1, threadStats.size());
+            assertEquals((int) Math.ceil(getSegmentCount(INDEX_NAME) / 5.0), threadStats.size());
```
Comment (reta) [inline server/src/internalClusterTest/java/org/opensearch/action/admin/cluster/node/tasks/ConcurrentSearchTasksIT.java:113]: Changes related to task executor for concurrent search (the current thread does work as well)

**Comment by reta** [comment]
@msfroh could you please take a look as well? thank you!

**Review by msfroh — APPROVED** [review]
Going through, I don't see any changes that seem unreasonable.

In theory, since `CompositeCodec` is guarded behind an experimental feature flag and hasn't shipped the search side of things yet, we may not need to keep `Composite99Codec` around, depending on our feelings on backwards compatibility for experimental features.

**Comment by bharath-techie** [comment]
I guess I missed this , @msfroh yes that certainly makes sense. We can completely drop Composite99Codec since its not shopped yet.

Thanks @reta will refactor and include the above changes.

**Pull request merged**